### PR TITLE
Do not suggest duplicate bounds

### LINKED ISSUE #68205: Using associated constant as array length: without (funny error) and with (ICE) const generics
Given the following test and nightly 2020-01-12:
```
#![allow(unused)]
#![feature(const_generics)]

trait Child {
    const N: usize;
}

struct C1;
impl Child for C1 {
    const N: usize = 2;
}

struct Main<C: Child> {
    c: C
}

impl<C: Child> Main<C> {
    fn test() -> [u8; C::N] {
        [0; C::N]
    }
}
```

I get that this is not supposed to work without const generics, but the error message suggestion is very funny:
```
error[E0599]: no associated item named `N` found for type parameter `C` in the current scope
  --> src/main.rs:22:27
   |
22 |     fn test2() -> [u8; C::N] {
   |                           ^ associated item not found in `C`
   |
   = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `N`, perhaps you need to restrict type parameter `C` with it:
   |
17 | impl<C: Child + Child> Main<C> {
   |      ^^^^^^^^^^
```

With const generics:
```
warning: Error finalizing incremental compilation session directory `/home/gbr/devel/rust-test/target/debug/incremental/rust_test-xwudz02f6bay/s-fjpbjlpyxj-1ihvwt0-working`: No such file or directory (os error 2)

error: internal compiler error: cat_expr Errd
  --> src/main.rs:22:30
   |
22 |       fn test2() -> [u8; C::N] {
   |  ______________________________^
23 | |         [0; C::N]
24 | |     }
   | |_____^

error: internal compiler error: cat_expr Errd
  --> src/main.rs:23:9
   |
23 |         [0; C::N]
   |         ^^^^^^^^^

error: internal compiler error: PromoteTemps: MIR had errors
  --> src/main.rs:22:5
   |
22 | /     fn test2() -> [u8; C::N] {
23 | |         [0; C::N]
24 | |     }
   | |_____^

error: internal compiler error: broken MIR in DefId(0:16 ~ rust_test[3df1]::{{impl}}[1]::test2[0]) ("return type"): bad type [type error]
  --> src/main.rs:22:5
   |
22 | /     fn test2() -> [u8; C::N] {
23 | |         [0; C::N]
24 | |     }
   | |_____^

error: internal compiler error: broken MIR in DefId(0:16 ~ rust_test[3df1]::{{impl}}[1]::test2[0]) (LocalDecl { mutability: Mut, local_info: Other, internal: false, is_block_tail: None, ty: [type error], user_ty: UserTypeProjections { contents: [] }, source_info: SourceInfo { span: src/main.rs:22:5: 24:6, scope: scope[0] } }): bad type [type error]
  --> src/main.rs:22:5
   |
22 | /     fn test2() -> [u8; C::N] {
23 | |         [0; C::N]
24 | |     }
   | |_____^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:346:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (3ebcfa145 2020-01-12) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden
```

### LINKED ISSUE #68695: Confusing diagnostics when using associated const as an array size
```rust
trait Adapter {
    const LINKS: usize;
}

struct Foo<A: Adapter> {
    adapter: A,
    links: [u32; A::LINKS],
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=39e7693cf8a87f30e40a4d280c312e6a)
```
error[E0599]: no associated item named `LINKS` found for type `A` in the current scope
 --> src/lib.rs:7:21
  |
7 |     links: [u32; A::LINKS],
  |                     ^^^^^ associated item not found in `A`
  |
  = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `LINKS`, perhaps you need to restrict type parameter `A` with it:
  |
5 | struct Foo<A: Adapter + Adapter> {
  |            ^^^^^^^^^^^^
```

At the same time on nightly with `#![feature(const_generics)]` error description is much better:
```rust
#![feature(const_generics)]
trait Adapter {
    const LINKS: usize;
}

struct Foo<A: Adapter> {
    adapter: A,
    links: [u32; A::LINKS],
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5af78935c897d1c7c5395ac820a03ae8)

```
error: constant expression depends on a generic parameter
 --> src/lib.rs:8:5
  |
8 |     links: [u32; A::LINKS],
  |     ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this may fail depending on what value the parameter takes
```

### PR DESCRIPTION
**Pull request by JohnTitor**
Fixes #68205
Fixes #68695

r? @estebank

### CODE REVIEW

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -853,26 +853,30 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                                err.span_suggestions(
-                                    sp,
-                                    &message(format!(
-                                        "restrict type parameter `{}` with",
-                                        param.name.ident(),
-                                    )),
-                                    candidates.iter().map(|t| {
-                                        format!(
-                                            "{}{} {}{}",
+                                let trait_def_ids: Vec<DefId> = param
```
Comment (Centril) [inline src/librustc_typeck/check/method/suggest.rs]:
```suggestion
                                let trait_def_ids: FxHashMap<DefId> = param
```

Comment (JohnTitor) [reply src/librustc_typeck/check/method/suggest.rs]: You mean `FxHashSet`?

Comment (Centril) [reply src/librustc_typeck/check/method/suggest.rs]: Yeah :)

Comment (JohnTitor) [reply src/librustc_typeck/check/method/suggest.rs]: Thanks, fixed.

**Pull request merged**